### PR TITLE
don't try to run lief on static libraries

### LINKED ISSUE #4787: Unexpected `binary is None` in `liefldd.py`
### Checklist

- [X] I added a descriptive title
- [X] I searched open reports and couldn't find a duplicate

### What happened?

With the last conda-smithy release, conda-forge is now back to using lief 0.12.3, and failing in https://github.com/conda-forge/compiler-rt-feedstock/pull/61 with:
```
  File "/Users/runner/miniforge3/lib/python3.10/site-packages/conda_build/os_utils/liefldd.py", line 406, in inspect_linkages_lief
    if binary.format == lief.EXE_FORMATS.ELF:
AttributeError: 'NoneType' object has no attribute 'format'
```

### Conda Info

```shell
conda-forge CI for osx-64 / osx-arm64
```

### Conda Config

```shell
conda-forge CI for osx-64 / osx-arm64
```

### Conda list

host env:
```shell
environment location: /Users/runner/miniforge3/conda-bld/compiler-rt-packages_1677355103841/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol


The following NEW packages will be INSTALLED:

    clang:           16.0.0.rc3-h9d310f0_0         conda-forge/label/llvm_rc
    clang-16:        16.0.0.rc3-default_h5b99a12_0 conda-forge/label/llvm_rc
    clang-format:    16.0.0.rc3-default_h5b99a12_0 conda-forge/label/llvm_rc
    clang-format-16: 16.0.0.rc3-default_h5b99a12_0 conda-forge/label/llvm_rc
    clang-tools:     16.0.0.rc3-default_h5b99a12_0 conda-forge/label/llvm_rc
    clangdev:        16.0.0.rc3-default_h5b99a12_0 conda-forge/label/llvm_rc
    clangxx:         16.0.0.rc3-default_h5b99a12_0 conda-forge/label/llvm_rc
    icu:             70.1-h96cf925_0               conda-forge              
    libclang:        16.0.0.rc3-default_h5b99a12_0 conda-forge/label/llvm_rc
    libclang-cpp:    16.0.0.rc3-default_h5b99a12_0 conda-forge/label/llvm_rc
    libclang-cpp16:  16.0.0.rc3-default_h5b99a12_0 conda-forge/label/llvm_rc
    libclang13:      16.0.0.rc3-default_h4e9855e_0 conda-forge/label/llvm_rc
    libcxx:          14.0.6-hccf4f1f_0             conda-forge              
    libiconv:        1.17-hac89ed1_0               conda-forge              
    libllvm16:       16.0.0.rc3-h8456b07_1         conda-forge/label/llvm_rc
    libxml2:         2.10.3-hb9e07b5_0             conda-forge              
    libzlib:         1.2.13-hfd90126_4             conda-forge              
    llvm-tools:      16.0.0.rc3-h8456b07_1         conda-forge/label/llvm_rc
    llvmdev:         16.0.0.rc3-h8456b07_1         conda-forge/label/llvm_rc
    xz:              5.2.6-h775f41a_0              conda-forge              
    zlib:            1.2.13-hfd90126_4             conda-forge              
    zstd:            1.5.2-hbc0c0cd_6              conda-forge 
```
build env:
```
  environment location: /Users/runner/miniforge3/conda-bld/compiler-rt-packages_1677355103841/_build_env

The following NEW packages will be INSTALLED:

    bzip2:                  1.0.8-h0d85af4_4          conda-forge
    c-ares:                 1.18.1-h0d85af4_0         conda-forge
    ca-certificates:        2022.12.7-h033912b_0      conda-forge
    clang_bootstrap_osx-64: 14.0.6-h6ddedbb_4         conda-forge
    cmake:                  3.25.2-h4032537_0         conda-forge
    expat:                  2.5.0-hf0c8a7f_0          conda-forge
    krb5:                   1.20.1-h049b76e_0         conda-forge
    libcurl:                7.88.1-h6df9250_0         conda-forge
    libcxx:                 14.0.6-hccf4f1f_0         conda-forge
    libedit:                3.1.20191231-h0678c8f_2   conda-forge
    libev:                  4.33-haf1e3a3_1           conda-forge
    libffi:                 3.4.2-h0d85af4_5          conda-forge
    libnghttp2:             1.51.0-he2ab024_0         conda-forge
    libsqlite:              3.40.0-ha978bb4_0         conda-forge
    libssh2:                1.10.0-h47af595_3         conda-forge
    libuv:                  1.44.2-hac89ed1_0         conda-forge
    libzlib:                1.2.13-hfd90126_4         conda-forge
    ncurses:                6.3-h96cf925_1            conda-forge
    ninja:                  1.11.1-hb8565cd_0         conda-forge
    openssl:                3.0.8-hfd90126_0          conda-forge
    python:                 3.11.0-he7542f4_1_cpython conda-forge
    readline:               8.1.2-h3899abd_0          conda-forge
    rhash:                  1.4.3-hac89ed1_0          conda-forge
    tk:                     8.6.12-h5dbffcc_0         conda-forge
    tzdata:                 2022g-h191b570_0          conda-forge
    xz:                     5.2.6-h775f41a_0          conda-forge
    zlib:                   1.2.13-hfd90126_4         conda-forge
    zstd:                   1.5.2-hbc0c0cd_6          conda-forge
```

### Additional Context

Failing CI run [here](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=664408&view=results)

### PR DESCRIPTION
**Pull request by h-vetinari**
Fixes #4787

### CODE REVIEW

**Comment by h-vetinari** [comment]
Just signed the CLA, let me know if there's anything else left to do to unblock this aspect.

**Comment by 1player** [comment]
I tried this patch on my custom build of LLVM/compiler-rt on macOS and conda build just hung on a loop when it reached the packaging phase. Have you confirmed that it works for you?

**Comment by h-vetinari** [comment]
> Have you confirmed that it works for you?

No, so far I just followed the suggestion given in the issue (+ some educated guessing, e.g. what's going to be in `files`). It's not clear to me how to write tests for this, but I also don't know the conda-build code base.

On the other hand, I cannot see how this would hang more or differently with the patch than without. There's no recursion that I can see (`_collect_needed_dsos` is only called once in the whole codebase), and it just means we do strictly less work (by `continue`-ing a bunch of items in a loop; an operation that's also done in another case just below).

**Comment by 1player** [comment]
Yes, you are right once again :) It just takes an enormous amount of time to do the post-compilation checks on macOS. Takes literally as long as the entire compilation of LLVM, but seems to complete eventually, so looks like this fix is the way to go. Thanks.

**Comment by 1player** [comment]
I'm not sure this is working correctly... There are some .a files generated by LLVM's compiler-rt that are not copied in the final tarball, and during compilation of LLVM I see a ton of "is not an ELF" and "Relocation \#1 of __text seems corrupted".

I also see a "WARNING :: Failed to get_static_lib_exports" on the archive file that was previously breaking conda-build (lib/clang/../lib/libclang_rt.builtins_i386_osx.a), and with this PR this file is skipped altogether and not included in the package tree, which is incorrect behaviour.

**Comment by h-vetinari** [comment]
> I'm not sure this is working correctly... There are some .a files generated by LLVM's compiler-rt that are not copied to the final tarball, and during compilation I see a ton of "is not an ELF" and "Relocation #1 of __text seems corrupted" messages.

What you're seeing is normal. Please have a look at the logs of the compiler-rt-feedstock before you conclude something must be amiss. The changes here are intentionally minimal, and will not affect which static libraries are packaged or not (that's the job of the recipe build scripts).

**Comment by 1player** [comment]
Be that as it may, there might have been a user error the first time, but I tried again this patch on a fresh install of macOS and miniconda3, and I got this error during packaging:

```
Traceback (most recent call last):
  File "/Users/user/miniconda3/bin/conda-build", line 11, in <module>
    sys.exit(main())
  File "/Users/user/miniconda3/lib/python3.10/site-packages/conda_build/cli/main_build.py", line 495, in main
    execute(sys.argv[1:])
  File "/Users/user/miniconda3/lib/python3.10/site-packages/conda_build/cli/main_build.py", line 475, in execute
    outputs = api.build(
  File "/Users/user/miniconda3/lib/python3.10/site-packages/conda_build/api.py", line 180, in build
    return build_tree(
  File "/Users/user/miniconda3/lib/python3.10/site-packages/conda_build/build.py", line 3097, in build_tree
    packages_from_this = build(metadata, stats,
  File "/Users/user/miniconda3/lib/python3.10/site-packages/conda_build/build.py", line 2369, in build
    newly_built_packages = bundlers[pkg_type](output_d, m, env, stats)
  File "/Users/user/miniconda3/lib/python3.10/site-packages/conda_build/build.py", line 1658, in bundle_conda
    files = post_process_files(metadata, initial_files)
  File "/Users/user/miniconda3/lib/python3.10/site-packages/conda_build/build.py", line 1504, in post_process_files
    post_build(m, new_files, build_python=python)
  File "/Users/user/miniconda3/lib/python3.10/site-packages/conda_build/post.py", line 1321, in post_build
    check_overlinking(m, files, host_prefix)
  File "/Users/user/miniconda3/lib/python3.10/site-packages/conda_build/post.py", line 1227, in check_overlinking
    return check_overlinking_impl(m.get_value('package/name'),
  File "/Users/user/miniconda3/lib/python3.10/site-packages/conda_build/post.py", line 1146, in check_overlinking_impl
    needed = needed_dsos_for_file[f]
KeyError: 'lib/clang/14.0.6/lib/darwin/libclang_rt.osx.a'
```

**Comment by 1player** [comment]
I was finally able to make this work by reverting this change, but instead updating the `codefile_class` function in https://github.com/conda/conda-build/blob/6751f09968f00ae175c5c9df941f0171f156984b/conda_build/os_utils/pyldd.py#L1012 to return None if the file ends in `.a`. 

Feel free to incorporate it into your PR if you feel it makes sense.

**Comment by beeankha** [comment]
@conda-bot check

**Comment by jezdez** [comment]
This requires a news file and if possible a test case

**Comment by h-vetinari** [comment]
Sorry this fell off the radar

> @1player: Feel free to incorporate it into your PR if you feel it makes sense.

I did that, but I'm flying blind here - the function is not documented and I don't know the internals here. But I'm glad you tested it, thanks a lot!

> @jezdez: This requires a news file and if possible a test case

Added a news file; not sure how a test case would look like, short of building one of the failing feedstocks.

**Comment by h-vetinari** [comment]
Gentle ping @beeankha @jezdez on how to proceed here

File: news/4787-fix-lief [hunk 1]
```diff
@@ -0,0 +1,19 @@
+### Enhancements
+
+* <news item>
+
+### Bug fixes
+
+* Fixed handling of static libraries with newer (py)lief versions.
```
Comment (beeankha) [inline news/4787-fix-lief:7]:
```suggestion
* Fixed handling of static libraries with newer (py)lief versions. (#4808)
```

**Comment by jezdez** [comment]
Closing in favor of https://github.com/conda/conda-build/pull/4900

**Pull request closed without merging**